File: 678. Valid Parenthesis String.go
func checkValidString(s string) bool {
    openCount, closeCount, length := 0, 0, len(s)
    for i := range(length) {
        if s[i] == '(' || s[i] == '*' {
            openCount += 1
        } else {
            openCount -= 1
        }
        if s[length-1-i] == ')' || s[length-1-i] == '*' {
            closeCount += 1
        } else {
            closeCount -= 1
        }
        if openCount < 0 || closeCount < 0 {
            return false
        }
    }
    return true
}

File: 678. Valid Parenthesis String.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        open_count, close_count = 0, 0
        length = len(s)
        for i in range(0, length):
            if s[i] == '(' or s[i] == '*':
                open_count += 1
            else:
                open_count -= 1
            if s[-i-1] == ')' or s[-i-1] == '*':
                close_count += 1
            else:
                close_count -= 1
            if open_count < 0 or close_count < 0:
                return False
        return True

